s3fs/core.py

```diff
@@ -461,12 +461,25 @@ def put_tags(self, path, tags):
         path: str
             Existing key to attach tags to
         tags: dict str, str
-            Tags to apply. Will over-write any existing tags.
+            Tags to apply.
+        mode:
+            One of 'o' or 'm'
+            'o': Will over-write any existing tags.
+            'm': Will merge in new tags with existing tags.  Incurs two remote calls.
         """
         bucket, key = split_path(path)
-        tag = {'TagSet': [
-                   {'Key': k, 'Value': v} for k, v in tags.items()]
-               }
+
+        if mode == 'm':
+            existing_tags = self.get_tags(path=path)
+            merged_tags = existing_tags.copy()
+            merged_tags.update(tags)
+            new_tags = [{'Key': k, 'Value': v} for k, v in merged_tags.items()]
```

Comment:
could be a little simpler? 
```
if mode == 'm':
    new_tags = [{'Key': k, 'Value': v} for k, v in list(tags.items()) + list(self.get_tags(path=path).items())]
```
Maybe you don't think that's simpler... in any case, I don't think the `copy()` line is necessary.

Reply:
I think you're right, we could get rid of copy.  I wasn't sure how something like you suggest would handle updating the values of existing keys.  I used "update" since it is compatible with python 2.7 -- as of python 3.5 it looks like there is simpler way to merge two dicts e.g. {**x, **y}.   I will add a test and change the raise to ValueError.